src/dataflow/src/render/join/delta_join.rs

```diff
@@ -306,17 +281,10 @@ where
                                     bound_inputs.push(*other);
                                 }
 
-                                // We must now apply `mfp` as it *should* have sufficient support for
-                                // full evaluation. Before we do this, we need to permute it to refer
-                                // to the correct physical column locations.
-                                mfp.permute(&column_map, column_map.len());
-                                // The `mfp` has access to all columns at this point, and could
-                                // plausibly be much simpler that a general `MapFilterProject`.
-                                // As of this writing, the `mfp` will only contain literals and
-                                // some projection to place them and potentially copy columns.
-                                // These should not error (if the literals are non-errors) and
-                                // could result in a simpler (non-erroring) operator.
-                                if !mfp.is_identity() {
+                                // We have completed the join building, but may have work remaining.
+                                // For example, we may have expressions not pushed down (e.g. literals)
+                                // and projections that could not be applied (e.g. column repetition).
+                                if let Some(mfp) = join_build_state.complete() {
```

Comment:
so it seems like the assert in `complete` that checks if equivalences is empty would be new behavior right? just want to double check my understanding

Reply:
Well, it's new in the sense that previously we would just ignore non-empty `equivalence` and produce incorrect answers. I think it counts as a non-recoverable error, and our philosophy there is to tear things down. But yes, different behavior in that error case, yes!